NDTensors/src/abstractarray/permutedims.jl

```diff
@@ -0,0 +1,29 @@
+## NOTICE!!: Here we are not importing Base.permutedims or Base.permutedims! but
+## are writing our own implementation. This allows us to 
+# NDTensors.permutedims
+function permutedims(M::AbstractArray, perm)
+  return permutedims(leaf_parenttype(M), M, perm)
+end
+
+# NDTensors.permutedims
+function permutedims(::Type{<:AbstractArray}, M, perm)
+  return Base.permutedims(M, perm)
+end
+
+# NDTensors.permutedims!
+function permutedims!(Mdest::AbstractArray, M::AbstractArray, perm)
+  return permutedims!(leaf_parenttype(Mdest), Mdest, leaf_parenttype(M), M, perm)
+end
+
+# NDTensors.permutedims!
+function permutedims!(::Type{<:AbstractArray}, Mdest, ::Type{<:AbstractArray}, M, perm)
+  return Base.permutedims!(Mdest, M, perm)
+end
+
+function permutedims!!(B::AbstractArray, A::AbstractArray, perm, f)
+  return permutedims!!(leaf_parenttype(B), B, leaf_parenttype(A), A, perm, f)
+end
+
+function permutedims!!(::Type{<:AbstractArray}, B, ::Type{<:AbstractArray}, A, perm, f)
+  return B .= f.(B, Base.permutedims(A, perm))
+end
```

Comment:
I think it would be better to organize this into two layers:
```julia
function permutedims!!(Bleaftype::Type{<:AbstractArray}, B, Aleaftype::Type{<:AbstractArray}, A, perm, f)
  permutedims!(Bleaftype, B, Aleaftype, A, f)
end

function permutedims!(::Type{<:AbstractArray}, B, ::Type{<:AbstractArray}, A, perm, f)
  return B .= f.(B, Base.permutedims(A, perm))
end
```
so then types could customize either one. In general, this is how we should organize maybe-mutating functions, where by default they basically just forward to a mutating function but then they could do something else if the type is immutable, like `UnallocatedZeros`.